core/templates/dev/head/pages/admin/roles_tab/RoleGraphDirective.js

```diff
@@ -0,0 +1,222 @@
+            $scope, $element, $timeout, $filter, StateGraphLayoutService,
+            MAX_NODES_PER_ROW, MAX_NODE_LABEL_LENGTH) {
+          // A rough upper bound for the width of a single letter, in pixels,
+          // to use as a scaling factor to determine the width of graph nodes.
+          // This is not an entirely accurate description because it also takes
+          // into account the horizontal whitespace between graph nodes.
+          var LETTER_WIDTH_IN_PIXELS = 10.5;
+          var HORIZONTAL_NODE_PROPERTIES = ['x0', 'width', 'xLabel'];
+          var VERTICAL_NODE_PROPERTIES = ['y0', 'height', 'yLabel'];
+          $scope.GRAPH_WIDTH = (
+            MAX_NODES_PER_ROW * MAX_NODE_LABEL_LENGTH * LETTER_WIDTH_IN_PIXELS
+            );
+
+          var getElementDimensions = function() {
+            return {
+              h: $element.height(),
+              w: $element.width()
+            };
+          };
+
+          $scope.getGraphHeightInPixels = function() {
+            return Math.max($scope.GRAPH_HEIGHT, 300);
+          };
+
+          $scope.drawGraph = function(
+              nodes, originalLinks, initStateId, finalStateIds) {
+            $scope.finalStateIds = finalStateIds;
+            var links = angular.copy(originalLinks);
+
+            var nodeData = StateGraphLayoutService.computeLayout(
+              nodes, links, initStateId, angular.copy(finalStateIds));
+
+            var maxDepth = 0;
+            for (var nodeId in nodeData) {
+              maxDepth = Math.max(maxDepth, nodeData[nodeId].depth);
+            }
+            $scope.GRAPH_HEIGHT = 70.0 * (maxDepth + 1);
+
+            // Change the position values in nodeData to use pixels.
+            for (var nodeId in nodeData) {
+              for (var i = 0; i < HORIZONTAL_NODE_PROPERTIES.length; i++) {
+                nodeData[nodeId][HORIZONTAL_NODE_PROPERTIES[i]] = (
+                  $scope.GRAPH_WIDTH *
+                  nodeData[nodeId][HORIZONTAL_NODE_PROPERTIES[i]]);
+                nodeData[nodeId][VERTICAL_NODE_PROPERTIES[i]] = (
+                  $scope.GRAPH_HEIGHT *
+                  nodeData[nodeId][VERTICAL_NODE_PROPERTIES[i]]);
+              }
+            }
+
+            // These constants correspond to the rectangle that, when clicked
+            // and dragged, translates the graph. Its height, width, and x and
+            // y offsets are set to arbitrary large values so that the
+            // draggable area extends beyond the graph.
+            $scope.VIEWPORT_WIDTH = Math.max(10000, $scope.GRAPH_WIDTH * 5);
+            $scope.VIEWPORT_HEIGHT = Math.max(10000, $scope.GRAPH_HEIGHT * 5);
+            $scope.VIEWPORT_X = -Math.max(1000, $scope.GRAPH_WIDTH * 2);
+            $scope.VIEWPORT_Y = -Math.max(1000, $scope.GRAPH_HEIGHT * 2);
+
+            var graphBounds = StateGraphLayoutService.getGraphBoundaries(
+              nodeData);
+
+            $scope.augmentedLinks = links.map(function(link) {
```

Comment:
can all the augmentedLinks stuff (including the for loop below) be pulled out into the StateGraphLayoutService?
then you can just have `$scope.augmentedLinks = StateGraphLayoutService.getAugmentedLinks(...)` or something